### PR TITLE
Magic Number

### LINKED ISSUE #767: Magic number: rule is not supported in the diktat
Need to think about the logic of detecting of Magic Numbers. Right now it is not supported in the diktat. 
May be we should even do the following: for the very long mathematical expressions calculate them and insert into the code.  

`val a = someObj + (1 + 2 + 3)` should trigger a warning that `(1 + 2 + 3)` should be moved to constant

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
   Magic number

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [ ] Added tests on fixers
* [x] Updated diktat-analysis.yml
* [x] Updated available-rules.md
* [x] closes #767

### CODE REVIEW

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
what rule do we have in a code style for this. Will you add? Or we have existing?

**Comment by kentr0w** [comment]
No, we didn't have a rule for that. So I added this rule to chapter 3

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt [hunk 1]
```diff
@@ -112,6 +112,7 @@ enum class Warnings(
     ENUMS_SEPARATED(true, "3.9.1", "enum is incorrectly formatted"),
     WHEN_WITHOUT_ELSE(true, "3.11.1", "each 'when' statement must have else at the end"),
     LONG_NUMERICAL_VALUES_SEPARATED(true, "3.14.2", "long numerical values should be separated with underscore"),
+    MAGIC_NUMBER(false, "3.14.3", "defining constants with clear names describing what the magic number means"),
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt:115]:
```suggestion
    MAGIC_NUMBER(false, "3.14.3", "avoid using magic numbers, instead define constants with clear names describing what the magic number means"),
```

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.psiUtil.isExtensionDeclaration
+import org.jetbrains.kotlin.psi.psiUtil.parents
+
+/**
+ * Rule for magic number
+ */
+class MagicNumberRule(configRules: List<RulesConfig>) : DiktatRule("magic-number", configRules, listOf(MAGIC_NUMBER)) {
+    override fun logic(node: ASTNode) {
+        val configuration = MagicNumberConfiguration(
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt:34]: You shouldn't create a configuration instance on every visitor invocation; extract it to a class property or at least move inside condition

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        /**
+         * @param param parameter from config
+         * @return value of parameter
+         */
+        @Suppress("UnsafeCallOnNullableType")
+        fun getParameter(param: String) = config[param]?.toBoolean() ?: mapConfiguration[param]!!
+    }
+
+    companion object {
+        val ignoreNumbersList = listOf("-1", "1", "0", "2")
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt:89]: We should parse these numbers as numbers when creating configuration instance

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            checkNumber(node, configuration)
+        }
+    }
+
+    @Suppress("ComplexMethod")
+    private fun checkNumber(node: ASTNode, configuration: MagicNumberConfiguration) {
+        val nodeText = node.treePrev?.let { if (it.elementType == OPERATION_REFERENCE && it.hasChildOfType(MINUS)) "-${node.text}" else node.text } ?: node.text
+        val isIgnoreNumber = configuration.ignoreNumbers.contains(nodeText)
+        val isHashFunction = node.parent({ it.elementType == FUN && it.isHashFun() }) != null &&
+                node.parents().find { it.elementType == PROPERTY } == null
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt:47]: What does this condition mean?

File: info/guide/guide-chapter-3.md [hunk 5]
```diff
@@ -817,6 +817,19 @@ val socialSecurityNumber = 999_99_9999L
 val hexBytes = 0xFF_EC_DE_5E
 val bytes = 0b11010010_01101001_10010100_10010010
 ```
+#### <a name="r3.14.3"></a> 3.14.3: Magic number
+Prefer defining constants with clear names describing what the magic number means.
+**Valid example**:
+```kotlin
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:823]: Please split the example into invalid and then valid

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt (hunk 3 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt:89]: But there can be a long(`1L`) or float(`1.0f`) type, which we should also ignore. I thought to store all types of numbers in one list as String.

Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt:89]: Won't they have a different node element type (like integer constant vs double constant or something)? Also, you are then comparing these numbers by text, does it work then? Do you have tests for `1.0f`?

Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/MagicNumberRule.kt:89]: Yes, there are two types: `INTEGER_CONSTANT` and `FLOAT_CONSTANT` and I check it in `logic()`. I a little changed logic, but yes, I had test for this.

File: info/guide/guide-chapter-3.md [hunk 6]
```diff
@@ -819,6 +819,26 @@ val socialSecurityNumber = 999_99_9999L
...
+class Person() {
+
+    fun isAdult(age: Int): Boolean = age >= majority
+
+    companion object {
+        const val majority = 18
+    }
+}
+```
+**InValid example**:
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:835]:
```suggestion
**Invalid example**:
```

File: info/guide/guide-chapter-3.md [hunk 7]
```diff
@@ -819,6 +819,26 @@ val socialSecurityNumber = 999_99_9999L
...
+```kotlin
+class Person() {
+    val oldAge = 65
```
Comment (petertrr) [inline info/guide/guide-chapter-3.md:838]: Please make both examples consistent: why do you need an additional field here? Having `18` inlined in invalid example and extracted to a constant in the valid one seems enough.

**Review by petertrr — APPROVED** [review]

**Review by akuleshov7 — APPROVED** [review]
Lgtm

**Pull request merged**